File: 백준/Silver/15652. N과 M （4）/N과 M （4）.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

    private static int N, M;
    private static int[] arr;
    private static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) throws IOException {
        
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        arr = new int[M];

        backTracking(1, 0);

        System.out.println(sb);
    }

    private static void backTracking(int start, int depth) {
        if (depth == M) {
            for (int val : arr) {
                sb.append(val).append(" ");
            }

            sb.append("\n");
            return;
        }

        for (int i = start; i <= N; i++) {
            arr[depth] = i;
            backTracking(i, depth + 1);
        }
    }
}

File: 백준/Silver/15652. N과 M （4）/N과 M （4）.py
import sys

input = sys.stdin.readline

N, M = map(int, input().split())
result = []

def backtracking(start):
    if len(result) == M:
        print(*result)
        return
    
    for i in range(start, N + 1):
        result.append(i)
        backtracking(i)
        result.pop()

backtracking(1)